src/lib/scheduler/abstract_task.hpp

```diff
@@ -133,10 +133,10 @@ class AbstractTask : public std::enable_shared_from_this<AbstractTask> {
    */
   void _on_predecessor_done();
 
-  TaskID _id = INVALID_TASK_ID;
-  NodeID _node_id = INVALID_NODE_ID;
-  SchedulePriority _priority;
-  bool _stealable;
+  std::atomic<TaskID> _id{INVALID_TASK_ID};
+  std::atomic<NodeID> _node_id = INVALID_NODE_ID;
+  std::atomic<SchedulePriority> _priority;
+  std::atomic_bool _stealable;
```

Comment:
I'm not sure if we actually need all of these to be atomic. `_stealable` for example is set in the constructor, and then never written, only read. Do we need an atomic there?

Reply:
I wouldn't say so: https://stackoverflow.com/questions/35598311/c-constructor-thread-safety

Reply:
Consider me a MT noob. 
Which mechanism *guarantees* that the write operation in the constructor is ordered before the read operation in another thread? 
Two weeks ago I would have argued the same as you: Obviously the constructor happens before anything else, so obviously its actions will be visible to another thread, which after all can only read the data after the constructor finished. Now I am not so sure anymore. The data written in the constructor needs to be flushed into *main memory* in order to be readable from another CPU.

Reply:
It's possible I am being totally stupid here. Please help me.

Reply:
> The data written in the constructor needs to be flushed into main memory in order to be readable from another CPU.

Negative. It is sufficient if it is in one of the CPU caches. The coherency protocols will ensure that it can be read from other threads.

The issue with `_id` is that `set_id` is called in `NodeQueueScheduler::schedule`, but there is no strict requirement for `set_id` to finish before `queue->push(task, ...)` is called. That is because push operates on a `shared_ptr` that is not "associated with [the] argument expression" `_id`. Thus, within one thread, it would be fine if `_id` is set later. With two threads, the second thread may take the task from the queue before `_id` is set and we have a data race.

This is different with the arguments set in the constructor. One of the rules [here](https://en.cppreference.com/w/cpp/language/eval_order) guarantees that the constructor is sequenced before anything happens with the object.

Take this with a grain of salt though - I can't claim to fully understand the memory model myself.

Reply:
Okay, seems like I have a bit of reading to do

Reply:
Reverted the atomicity of stealable and priority since they are only written to in the constructor

Reply:
I would also go into the implementation of the get/set methods of node and task id, and specify load/store explicitly with the correct [memory_order](https://en.cppreference.com/w/cpp/atomic/memory_order) (memory_order_acquire for atomic.load() in the get() methods, and memory_order_release for the atomic.store() in the set() methods if I read and understand that correctly). I know you probably just wanted a quick fix for the master and get over with it, and if you don't want to deal with that right now you can pass it to me, but I think if we "fix" it then we should do it correctly.

Also, the first half of the benchmarks (the one without your changes) were done last night, the next half I'll schedule for next night (on pella), so we should have them by tomorrow.

Reply:
I'd merge this one asap and let you improve it once you have the time. I'm trying to catch up on the whole memory model thing, but I wouldn't right now wanna deal with explicit memory orders.